### ISSUE TITLE
Doubt regarding the rotateStringArray, shuffleStringArray, stringArrayEncoding

### ISSUE THREAD

**Issue by gaurav-cointab** [javascript-obfuscator/javascript-obfuscator#724]
**Doubt regarding rotateStringArray, shuffleStringArray**
When an array item is accessed at runtime, which of these shifts and rotates the string array items?
And which one of them will just jumble the string array items so their ordering is not guessable?
The property which shifts and rotates the string array items will be expensive, if the code base is very very large and has a lot of string literals, and when we have the string split chunk length as 3,4?

**Doubt regarding stringArrayEncoding**
When this value is set as base64 or rc, then it will decode the array item at runtime, so they will always be decoded when those items are accessed right?
Is there a way to encode them, but decode them only once, and then use the decoded values continuously at runtime?

**Comment by sanex3339** [comment]
> When an array item is accessed at runtime, which of these shifts and rotates the string array items?
> And which one of them will just jumble the string array items so their ordering is not guessable?
> The property which shifts and rotates the string array items will be expensive, if the code base is very very large and has a lot of string literals, and when we have the string split chunk length as 3,4?
Main goal of `rotateStringArray` option is to make harder process when someone cannot run the code. So all items are rotated. At runtime all items are rotated back to their original places at the start of code evaluation. So, in the obfuscated code string array has different elements order then at the runtime.
Main goal of `shuffleStringArray` is the same: make harder process when someone cannot run the code. But with this option all items are randomly shuffled both at the obfuscated code and during runtime.

> When this value is set as base64 or rc, then it will decode the array item at runtime, so they will always be decoded when those items are accessed right?

No. For the performance reasons all values will be decoded during the first call of each value and then cached.

**Comment by sanex3339** [comment]
Closed. Feel free to write here or reopen a new issue if you have additional questions.
